File: ios/Classes/VideoPool.swift
import Foundation

/// Placeholder pool abstraction for index-aware AVPlayer reuse.
final class VideoPool {
  private let maxPoolSize: Int
  private var activeControllerIds: [Int] = []

  init(maxPoolSize: Int) {
    self.maxPoolSize = maxPoolSize
  }

  func markActive(_ controllerId: Int) {
    activeControllerIds.removeAll(where: { $0 == controllerId })
    activeControllerIds.append(controllerId)
    trimToPoolSize()
  }

  func markReleased(_ controllerId: Int) {
    activeControllerIds.removeAll(where: { $0 == controllerId })
  }

  func clear() {
    activeControllerIds.removeAll()
  }

  private func trimToPoolSize() {
    if activeControllerIds.count <= maxPoolSize {
      return
    }

    activeControllerIds.removeFirst(activeControllerIds.count - maxPoolSize)
  }
}

File: android/src/main/kotlin/com/example/native_feed_player/VideoPool.kt
package com.example.native_feed_player

/**
 * Placeholder pool abstraction for mapping feed index -> player instances.
 */
internal class VideoPool(
    private val maxPoolSize: Int
) {
    private val activeControllerIds = linkedSetOf<Int>()

    fun markActive(controllerId: Int) {
        activeControllerIds.add(controllerId)
        trimToPoolSize()
    }

    fun markReleased(controllerId: Int) {
        activeControllerIds.remove(controllerId)
    }

    fun clear() {
        activeControllerIds.clear()
    }

    private fun trimToPoolSize() {
        while (activeControllerIds.size > maxPoolSize) {
            val oldest = activeControllerIds.firstOrNull() ?: return
            activeControllerIds.remove(oldest)
        }
    }
}
